### PR TITLE
probe:fix: Fix startup automatic capturing

### PR DESCRIPTION
**Pull request by vperelmax**
Startup automatic capturing based on configuration
didnt work.
This patch is its fix.

### CODE REVIEW

File: tests/startup_capture_test.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+			}
+			if !found {
+				return errors.New("We should found the Hops nodes in the TrackingID nodes")
+			}
+
+			return nil
+		}},
+	}
+
+	RunStartupTest(t, test)
```
Comment (safchain) [inline tests/startup_capture_test.go]: Why copying the original RunTest and not using it ? did you do any changes, any feature is missing ?

Comment (vperelmax) [reply tests/startup_capture_test.go]: yes. they use list of captures provided by the test caller and create via client new captures. I dont want test to create new captures, but see that analyzer uses the pre-configured.
Of cause I can duplicate less code, but for tests I dont see if it is really needed

Comment (safchain) [reply tests/startup_capture_test.go]: Why not providing an empty list so that no captures will be created

Comment (vperelmax) [reply tests/startup_capture_test.go]: But then it also doesn't get the needed captures. I tried to do it. It doesnt work

File: tests/startup_capture_test.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+				t.Errorf("Failed to replay test: %s, graph: %s, flows: %s", err.Error(), checkContext.getWholeGraph(t, checkContext.time), checkContext.getAllFlows(t, checkContext.time))
+			}
+		}
+	}
+}
+
+func TestStartupCaptureNew(t *testing.T) {
+	test := &Test{
+		setupCmds: []helper.Cmd{
+			{"ovs-vsctl add-br br-startup", true},
```
Comment (safchain) [inline tests/startup_capture_test.go]: I think for this test you only need to create this bridge as you just want to test that the capture is started

File: tests/startup_capture_test.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+			{"ip netns exec fh-vm1 ip link set fh-intf1 up", true},
+
+			{"ovs-vsctl add-port br-startup fh-intf2 -- set interface fh-intf2 type=internal", true},
+			{"ip netns add fh-vm2", true},
+			{"ip link set fh-intf2 netns fh-vm2", true},
+			{"ip netns exec fh-vm2 ip address add 169.254.33.34/24 dev fh-intf2", true},
+			{"ip netns exec fh-vm2 ip link set fh-intf2 up", true},
+		},
+
+		settleFunction: func(c *TestContext) error {
```
Comment (safchain) [inline tests/startup_capture_test.go]: no need to have this

File: tests/startup_capture_test.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+			gremlin = `G.V().Has("Name", "fh-vm2").Out().Has("Name", "fh-intf2")`
+			nodes, err = c.gh.GetNodes(gremlin)
+			if err != nil || len(nodes) == 0 {
+				return errors.New("fh-intf2 not found in the expected namespace")
+			}
+
+			return nil
+		},
+
+		setupFunction: func(c *TestContext) error {
```
Comment (safchain) [inline tests/startup_capture_test.go]: no need to have this too

File: tests/startup_capture_test.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+		},
+
+		// since the agent update ticker is about 10 sec according to the configuration
+		// we should wait 11 sec to have the first update and the MetricRange filled
+		checks: []CheckFunction{func(c *CheckContext) error {
+			prefix := "g"
+			if !c.time.IsZero() {
+				prefix += fmt.Sprintf(".Context(%d)", common.UnixMillis(c.time))
+			}
+			prefix += `.V().Has("Name", "br-startup", "Type", "ovsbridge")`
```
Comment (safchain) [inline tests/startup_capture_test.go]: I would rather test that this node has Capture.ID field and that should be enough for this test

File: tests/startup_capture_test.go (hunk 1 above)
Comment (safchain) [reply tests/startup_capture_test.go]: So it seems to me that it is more related to your patch, the capture should start without changing anything, the analyzer/agent are not started by RunTest but by init(). I'll have a look

**Comment by vperelmax** [comment]
I have rewritten the whole test. Later today will upload it.

**Comment by vperelmax** [comment]
I checked the tests locally - everything passes :( I think need some help in re-running them

**Comment by nplanel** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by lebauce** [comment]
I would prefer to make it more standard:

```
diff --git a/tests/startup_capture_test.go b/tests/startup_capture_test.go
index c5b8afc8..40626ae4 100644
--- a/tests/startup_capture_test.go
+++ b/tests/startup_capture_test.go
@@ -1,17 +1,19 @@
 package tests
 
 import (
-        gclient "github.com/skydive-project/skydive/api/client"
+        "fmt"
+        "testing"
+
         "github.com/skydive-project/skydive/api/types"
         "github.com/skydive-project/skydive/common"
         "github.com/skydive-project/skydive/config"
-        shttp "github.com/skydive-project/skydive/http"
         "github.com/skydive-project/skydive/tests/helper"
-        "testing"
 )
 
 func TestStartupCapture(t *testing.T) {
         test := &Test{
+                mode: OneShot,
+
                 setupCmds: []helper.Cmd{
                         {"ovs-vsctl add-br br-startup", true},
                         {"ovs-vsctl add-port br-startup startup-intf2 -- set interface startup-intf2 type=internal", true},
@@ -25,60 +27,59 @@ func TestStartupCapture(t *testing.T) {
                         {"ip netns del startup-vm2", true},
                         {"ovs-vsctl del-br br-startup", true},
                 },
-        }
 
-        helper.ExecCmds(t, test.setupCmds...)
+                checks: []CheckFunction{func(c *CheckContext) error {
+                        //Get config captures
+                        var captures map[string]types.Capture
+                        if err := c.client.List("capture", &captures); err != nil {
+                                return err
+                        }
 
-        client, err := gclient.NewCrudClientFromConfig(&shttp.AuthenticationOpts{})
-        if err != nil {
-                t.Errorf("Failed to create client: %s", err.Error())
-        }
+                        gremlin := config.GetString("analyzer.startup.capture_gremlin")
+                        if gremlin == "" {
+                                return fmt.Errorf("No startup capture is defined in config!")
+                        }
 
-        //Get config captures
-        var captures map[string]types.Capture
-        if err := client.List("capture", &captures); err != nil {
-                t.Errorf(err.Error())
-        }
+                        var capture *types.Capture
+                        for _, tc := range captures {
+                                if gremlin == tc.GremlinQuery {
+                                        capture = types.NewCapture(tc.GremlinQuery, tc.BPFFilter)
+                                        break
+                                }
+                        }
 
-        gremlin := config.GetString("analyzer.startup.capture_gremlin")
-        if gremlin == "" {
-                t.Errorf("No startup capture is defined in config!")
-        }
-        var myCapture *types.Capture
-        for _, tc := range captures {
-                if gremlin == tc.GremlinQuery {
-                        myCapture = types.NewCapture(tc.GremlinQuery, tc.BPFFilter)
-                        break
-                }
-        }
-        if myCapture == nil {
-                t.Errorf("Preconfigured capture wasn't created!")
-        }
+                        if capture == nil {
+                                return fmt.Errorf("Preconfigured capture wasn't created!")
+                        }
 
-        gh := gclient.NewGremlinQueryHelper(&shttp.AuthenticationOpts{})
+                        nodes, err := c.gh.GetNodes(capture.GremlinQuery)
+                        if err != nil {
+                                return err
+                        }
 
-        nodes, err := gh.GetNodes(myCapture.GremlinQuery)
-        if err != nil {
-                t.Errorf(err.Error())
-        }
+                        if len(nodes) == 0 {
+                                return fmt.Errorf("No node matching capture %s", capture.GremlinQuery)
+                        }
 
-        if len(nodes) == 0 {
-                t.Errorf("No node matching capture %s", myCapture.GremlinQuery)
-        }
+                        for _, node := range nodes {
+                                tp, err := node.GetFieldString("Type")
+                                if err != nil || !common.IsCaptureAllowed(tp) {
+                                        continue
+                                }
+
+                                captureID, err := node.GetFieldString("Capture.ID")
+                                if err != nil {
+                                        return fmt.Errorf("Node %+v matched the capture but capture is not enabled", node)
+                                }
+
+                                if captureID != capture.ID() {
+                                        return fmt.Errorf("Node %s matches multiple captures - %s", node.ID, node.String())
+                                }
+                        }
 
-        for _, node := range nodes {
-                tp, err := node.GetFieldString("Type")
-                if err != nil || !common.IsCaptureAllowed(tp) {
-                        continue
-                }
-                captureID, err := node.GetFieldString("Capture.ID")
-                if err != nil {
-                        t.Errorf("Node %+v matched the capture but capture is not enabled", node)
-                }
-                if captureID != myCapture.ID() {
-                        t.Errorf("Node %s matches multiple captures - %s", node.ID, node.String())
-                }
+                        return nil
+                }},
         }
 
-        helper.ExecCmds(t, test.tearDownCmds...)
+        RunTest(t, test)
 }
```

**Comment by vperelmax** [comment]
I tried , locally it works :(

**Comment by vperelmax** [comment]
added more standard test

**Comment by lebauce** [comment]
@vperelmax Many CI issues should now be fixed. Can you please rebase on master ?

**Comment by vperelmax** [comment]
Hi, yes, sure.
Rebasing

**Comment by vperelmax** [comment]
Did rebase.

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by lebauce** [comment]
@vperelmax Thanks

**Pull request merged**